clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
For the `TimeoutException` case, it actually seems like an improvement to reduce the logging to a single warning instead of multiple (if the error message can be improved, that's even better). Are there cases where we rely exclusively on the warning and get no exception?

@rajinisivaram also mentioned `manually assigning a partition for consume` as the second case where the logging could be useful. What's the thinking for that one?